### ISSUE TITLE
Helper code for Lab 3

### ISSUE THREAD

**Issue by pxaris** [slp-ntua/slp-labs#124]
Issue που θα χρησιμοποιηθεί κατά τη διάρκεια του Εργαστηρίου (στις 2/6/23) για τη σημείωση βοηθητικού κώδικα, οδηγιών κ.α.

**Comment by pxaris** [comment]
### 0

- Prep-Lab: https://helios.ntua.gr/mod/assign/view.php?id=35806
- Lab: https://helios.ntua.gr/mod/assign/view.php?id=36122
- repo: https://github.com/slp-ntua/slp-labs/tree/master/lab3

**Comment by angeloskanatas** [comment]
### 1
```
# Convert data labels from strings to integers

from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()

label_encoder.fit(y_train)

y_train = label_encoder.transform(y_train)
y_test = label_encoder.transform(y_test)
n_classes = label_encoder.classes_.size
```

**Comment by Con-Hal** [comment]
###2
'''python
def __init__(self, X, y, word2idx):
        """
        In the initialization of the dataset we will have to assign the
        input values to the corresponding class attributes
        and preprocess the text samples

        - Store all meaningful arguments to the constructor here for debugging
          and for usage in other methods
        - Do most of the heavy-lifting like preprocessing the dataset here

        Args:
            X (list): List of training samples
            y (list): List of training labels
            word2idx (dict): a dictionary which maps words to indexes
        """
        
        self.data = X
        # Tokenize the sentences using nltk tokenizer
        self.tokenized_data = [word_tokenize(sentence) for sentence in self.data]
        self.labels = y
        self.word2idx = word2idx
        """

**Comment by Con-Hal** [comment]
### 3
'''
python

def __getitem__(self, index):
        """
        Returns the _transformed_ item from the dataset

        Args:
            index (int):

        Returns:
            (tuple):
                * example (ndarray): vector representation of a training example
                * label (int): the class label
                * length (int): the length (tokens) of the sentence

        Examples:
            For an `index` where:
            ::
                self.data[index] = ['this', 'is', 'really', 'simple']
                self.labels[index] = "neutral"

            the function will have to return something like:
            ::
                example = [  533  3908  1387   649   0     0     0     0]
                label = 1
                length = 4
        """
        example = self.tokenized_data[index]
        # map tokens to ids according to word2idx
        example = [self.word2idx.get(token, self.word2idx['<unk>']) for token in example]
        label = self.labels[index]
        length = len(example) 

        # zero padding using the maximum length from initialization
        # or truncation in case a larger example is found
        if length < self.max_length:
            example += [0] * (self.max_length - length)
        else:
            example = example[:self.max_length]
        
        return np.array(example), label, length
'''

**Comment by pxaris** [comment]
### 4, 5, 6
```python
class BaselineDNN(nn.Module):
    """
    1. We embed the words in the input texts using an embedding layer
    2. We compute the min, mean, max of the word embeddings in each sample
       and use it as the feature representation of the sequence.
    4. We project with a linear layer the representation
       to the number of classes.ngth)
    """

    def __init__(self, output_size, embeddings, trainable_emb=False):
        """

        Args:
            output_size(int): the number of classes
            embeddings(bool):  the 2D matrix with the pretrained embeddings
            trainable_emb(bool): train (finetune) or freeze the weights
                the embedding layer
        """

        super(BaselineDNN, self).__init__()

        # EX4
        # 1 - define the embedding layer      
        num_embeddings = len(embeddings) 
        dim = len(embeddings[0])
        self.embeddings = nn.Embedding(num_embeddings, dim)
        self.output_size = output_size

        # 2 - initialize the weights of our Embedding layer
        # from the pretrained word embeddings
        # 3 - define if the embedding layer will be frozen or finetuned
        if not trainable_emb:
            self.embeddings = self.embeddings.from_pretrained(torch.Tensor(embeddings), freeze = True)

        # 4 - define a non-linear transformation of the representations
        # EX5
        self.linear = nn.Linear(dim, 1000)
        self.relu = nn.ReLU()

        # 5 - define the final Linear layer which maps
        # the representations to the classes
        # EX5
        self.output = nn.Linear(1000, output_size)

    def forward(self, x, lengths):
        """
        This is the heart of the model.
        This function, defines how the data passes through the network.

        Returns: the logits for each class

        """

        # 1 - embed the words, using the embedding layer
        # EX6
        embeddings = self.embeddings(x)

        # 2 - construct a sentence representation out of the word embeddings
        # EX6
        # calculate the means
        representations = torch.sum(embeddings, dim=1)
        for i in range(lengths.shape[0]): # necessary to skip zeros in mean calculation
            representations[i] = representations[i] / lengths[i]
        
        # 3 - transform the representations to new ones.
        # EX6
        representations = self.relu(self.linear(representations))

        # 4 - project the representations to classes using a linear layer
        # EX6
        logits = self.output(representations)

        return logits
```

**Comment by Con-Hal** [comment]
### 7,8

'''
python

#############################################################################
# Training Pipeline
#############################################################################
# lists to acummulate train and test losses 
TRAIN_LOSS = []
TEST_LOSS = []
for epoch in range(1, EPOCHS + 1):
    # train the model for one epoch
    train_dataset(epoch, train_loader, model, criterion, optimizer)

    # evaluate the performance of the model, on both data sets
    train_loss, (y_train_gold, y_train_pred) = eval_dataset(train_loader,
                                                            model,
                                                            criterion)

    test_loss, (y_test_gold, y_test_pred) = eval_dataset(test_loader,
                                                         model,
                                                         criterion)
    TRAIN_LOSS.append(train_loss)
    TEST_LOSS.append(test_loss)
    # concatenate the predictions and gold labels in lists.
    y_train_true = np.concatenate( y_train_gold, axis=0 )
    y_test_true = np.concatenate( y_test_gold, axis=0 )
    y_train_pred = np.concatenate( y_train_pred, axis=0 )
    y_test_pred = np.concatenate( y_test_pred, axis=0 )
    # compute metrics using sklearn functions

    print("Train loss:" , train_loss)
    print("Test loss:", test_loss)
    print("Train accuracy:" , accuracy_score(y_train_true, y_train_pred))
    print("Test accuracy:" , accuracy_score(y_test_true, y_test_pred))
    print("Train F1 score:", f1_score(y_train_true, y_train_pred, average='macro'))
    print("Test F1 score:", f1_score(y_test_true, y_test_pred, average='macro'))
    print("Tain Recall:", recall_score(y_train_true, y_train_pred, average='macro'))
    print("Test Recall:", recall_score(y_test_true, y_test_pred, average='macro'))
# plot training and validation loss curves
    
plt.plot(range(1, EPOCHS + 1), TRAIN_LOSS, label='Training Loss')
plt.plot(range(1, EPOCHS + 1), TEST_LOSS, label='Validation Loss')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.title('Training and Validation Loss')
plt.legend()
plt.show()
'''

**Comment by pxaris** [comment]
### 9
```python
def train_dataset(_epoch, dataloader, model, loss_function, optimizer):
    # IMPORTANT: switch to train mode
    # enable regularization layers, such as Dropout
    model.train()
    running_loss = 0.0

    # obtain the model's device ID
    device = next(model.parameters()).device

    for index, batch in enumerate(dataloader, 1):
        # get the inputs (batch)
        inputs, labels, lengths = batch

        # EX9
        # move the batch tensors to the right device
        inputs = inputs.to(device)
        labels = labels.to(device)
        lengths = lengths.to(device)

        # Step 1 - zero the gradients
        # Remember that PyTorch accumulates gradients.
        # We need to clear them out before each batch!
        model.zero_grad()

        # Step 2 - forward pass: y' = model(x)
        if model.__class__.__name__ in ['BaselineDNN', 'LSTM']:
            outputs = model(inputs, lengths)
        else:
            outputs = model(inputs)

        # fix label type for different loss func
        # print(loss_function.__dict__)
        # if loss_function.__func__.__name__ == 'BCEWithLogitsLoss':
        #     labels = torch.nn.functional.one_hot(labels, num_classes=2)

        # Step 3 - compute loss: L = loss_function(y, y')
        try:
            loss = loss_function(outputs, labels)
        except ValueError:
            # fix labels for 'BCEWithLogitsLoss' loss function
            bin_labels = torch.nn.functional.one_hot(labels.long(), num_classes=2)
            loss = loss_function(outputs, bin_labels.float())


        # Step 4 - backward pass: compute gradient wrt model parameters
        loss.backward()

        # Step 5 - update weights
        optimizer.step()

        running_loss += loss.data.item()

        # print statistics
        progress(loss=loss.data.item(),
                 epoch=_epoch,
                 batch=index,
                 batch_size=dataloader.batch_size,
                 dataset_size=len(dataloader.dataset))

    return running_loss / index
```

**Comment by pxaris** [comment]
## Main Lab

**Comment by pxaris** [comment]
### 2.1

```python
# main.py
from training import torch_train_val_split
# train_loader = DataLoader(train_set, batch_size=BATCH_SIZE, shuffle=True) # REMOVE
train_loader, val_loader = torch_train_val_split(train_set, BATCH_SIZE, BATCH_SIZE)
```

```python
# main.py
#############################################################################
# Training Pipeline
#############################################################################
save_path = f'{DATASET}_{model.__class__.__name__}.pth'
early_stopper = EarlyStopper(model, save_path, patience=5) 
for epoch in range(1, EPOCHS + 1):
    # train the model for one epoch
    train_dataset(epoch, train_loader, model, criterion, optimizer)

    # evaluate the performance of the model, on both data sets
    train_loss, (y_train_gold, y_train_pred) = eval_dataset(train_loader,
                                                            model,
                                                            criterion)
    
    # evaluate the performance of the model, on both data sets
    valid_loss, (y_valid_gold, y_valid_pred) = eval_dataset(val_loader,
                                                            model,
                                                            criterion)
    
    print(f"\n===== EPOCH {epoch} ========")
    print(f'\nTraining set\n{get_metrics_report(y_train_gold, y_train_pred)}')
    print(f'\nValidation set\n{get_metrics_report(y_valid_gold, y_valid_pred)}')

    if early_stopper.early_stop(valid_loss):
        print('Early Stopping was activated.')
        print(f'Epoch {epoch}/{EPOCHS}, Loss at training set: {train_loss}\n\tLoss at validation set: {valid_loss}')
        print('Training has been completed.\n')
        break
```

**Comment by pxaris** [comment]
### 2.2
```python
# models.py:LSTM
# 2EX2
class LSTM(nn.Module):
    ...
    
    def forward(self, x, lengths):       
        ...

        # pick the output of the lstm corresponding to the last word
        representations = torch.zeros(batch_size, self.representation_size).float()
        for i in range(lengths.shape[0]):
            last = lengths[i] - 1 if lengths[i] <= max_length else max_length - 1
            representations[i] = ht[i, last, :]
       
        logits = self.linear(representations)
        
        return logits
```
